File: JavaScript/0019. Remove Nth Node From End of List.js
// Given a linked list, remove the n-th node from the end of list and return its head.
//
// Example:
// Given linked list: 1->2->3->4->5, and n = 2.
// After removing the second node from the end, the linked list becomes 1->2->3->5.
//
// Note:
// Given n will always be valid.
//
// Follow up:
// Could you do this in one pass?

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */

/** Two pointers slow and fast */
// Time O(n)
// Space O(1)
const removeNthFromEnd = (head, n) => {
  const preHead = new ListNode(null); // for case n = 1, to remove the last node in the list to avoid slow.next is null in slow.next.next
  preHead.next = head;

  let slow = preHead;
  let fast = head;

  while (n--) {
    fast = fast.next;
  }

  while (fast) {
    slow = slow.next;
    fast = fast.next;
  }

  slow.next = slow.next.next;
  return preHead.next;
};

File: Python/0019. Remove Nth Node From End of List.py
# Given the head of a linked list, remove the nth node from the end of the list and return its head.
#
# Example 1:
#
# Input: head = [1,2,3,4,5], n = 2
# Output: [1,2,3,5]
#
# Example 2:
#
# Input: head = [1], n = 1
# Output: []
#
# Example 3:
#
# Input: head = [1,2], n = 1
# Output: [1]
#
# Constraints:
#
# The number of nodes in the list is sz.
# 1 <= sz <= 30
# 0 <= Node.val <= 100
# 1 <= n <= sz
#
# Follow up: Could you do this in one pass?


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

# Two pointers slow and fast
# Time O(n)
# Space O(1)
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        pre_head = ListNode(None)
        pre_head.next = head

        slow = pre_head
        fast = head

        # Move fast first so that the gap between slow and fast becomes n
        for _ in range(n):
            fast = fast.next

        while fast:
            slow = slow.next
            fast = fast.next

        # Remove the node
        slow.next = slow.next.next
        return pre_head.next
